### PR TITLE
Add partial ungetc support

### PR DESCRIPTION
**Pull request by take-cheeze**

### CODE REVIEW

File: test/stringio.rb [hunk 1]
```diff
@@ -206,6 +206,15 @@ def o.to_s; "baz"; end
   assert_equal nil, strio.getc
 end
 
+assert 'StringIO#ungetc' do
+  strio = StringIO.new("abc")
+  assert_raise(IOError) { strio.ungetc '' }
```
Comment (ksss) [inline test/stringio.rb]: I expect same behavior as CRuby.

```rb
strio = StringIO.new("abc")

strio.ungetc ''
#=> nil

strio.string
#=> "abc"
```

File: test/stringio.rb [hunk 2]
```diff
@@ -206,6 +206,15 @@ def o.to_s; "baz"; end
...
+  assert_raise(IOError) { strio.ungetc 'a' }
```
Comment (ksss) [inline test/stringio.rb:213]: ditto.

```rb
strio = StringIO.new("abc")

strio.ungetc 'a'
#=> nil

strio.string
#=> "aabc"
```

```rb
strio = StringIO.new("abc", "w")
strio.ungetc 'a'
#=> IOError
```

**Review by ksss — COMMENTED** [review]
Thank you for sending PR!

**Comment by take-cheeze** [comment]
I don't think ungetc should be compatible with CRuby because there is too many corner cases to be careful
The only motivation for this PR is https://github.com/take-cheeze/mruby-marshal/issues/31

**Comment by ksss** [comment]
If so, I think you should either remove that test case or give an error message that clearly states that the corner case is unimplemented.
Current implementation looks like the specification.

**Comment by ksss** [comment]
The current implementation ( 92c06329b833803ab3c3c2840981c60d90e4a9bf ) is probably not behaving as you would expect.

```
$ ./build/test/bin/mirb
mirb - Embeddable Interactive Ruby Shell

> s = StringIO.new('abc')
 => #<StringIO:0x7fd63ed183b0>
> s.string
 => "abc"
> s.ungetc 'x'
mruby-ext: ungetc not supported in beginning of stream (IOError)
> s.pos += 1
 => 1
> s.ungetc 'x'
Assertion failed: (((((((struct RString*)((string).value.p)))->flags & 8)) ? (((struct RStringEmbed*)(((struct RString*)((string).value.p))))->ary) : (((struct RString*)((string).value.p)))->as.heap.ptr)[ptr->pos] == *in_data), function stringio_ungetc, file /Users/ksss/src/github.com/ksss/mruby-stringio/src/stringio.c, line 663.
[1]    54409 abort      ./build/test/bin/mirb
```

**Comment by ksss** [comment]
@take-cheeze How about this? https://github.com/ksss/mruby-stringio/pull/8

**Comment by take-cheeze** [comment]
Pure ruby implementation seems to be slow but I'm only using it for test for you decide

**Comment by ksss** [comment]
Thanks!
